### PR TITLE
Stop sending the product's name to the upstream provider (KBR-5)

### PR DESCRIPTION
**Pull request by shelpuk**
Fixes [KBR-5](https://shelpuk.atlassian.net/browse/KBR-5). Priority 0 — a live breach of the product's core promise.

## Context for the Product Owner

Kitty Bridge exists so the upstream provider cannot tell it is there: you point Claude Code at your own coding plan and the provider sees ordinary agent traffic. One code path broke that promise on our own initiative.

When the bridge could not shrink a conversation enough to send it, it threw the conversation away and sent the provider a message beginning **`[Kitty Bridge: Unable to compact conversation …]`**. Any provider that reads request bodies — and a coding-plan provider has every reason to — could identify our users from that single string. The rest of the design goes to considerable lengths to leave nothing else to find.

Two further harms, both fixed here:

- **The message never reached the person it was written for.** It was addressed to the user but delivered to the provider, which simply replied to it as if it were a prompt. The user saw the model's reaction, never the explanation.
- **The advice in it was wrong.** It blamed an oversized system prompt and told the user to run `/clear`. Neither is reliably true (see below).

**After this change:** the bridge sends nothing upstream, and returns a clear error to the agent in whichever error format that agent speaks. The user reads the message. The provider learns nothing.

## The documented trigger was wrong — measured, not inferred

The ticket, register row M13, finding F3 and the corpus row in `TEST_SUITE.md` all said this fires when *"the system prompt alone exceeds the model's context window"* — reachable with a large `CLAUDE.md` or many MCP tools. Running the real function against six shapes:

| Shape | Fires? |
|---|---|
| Huge system prompt **+ an ordinary user turn** — *the documented trigger* | **No** |
| Over-threshold conversation with no ordinary message at all | Yes |
| Unpaired tool result as the only surviving turn (Claude Code shape) | **Yes** |
| Unpaired tool result as the only surviving turn (native Anthropic shape) | **Yes** |
| Unpaired tool result at a tight model-derived budget | **Yes** |
| The existing test fixture for this path | **No** |

The compaction fallback always keeps at least one real message, so a surviving user turn always prevents the failure. Only the orphan-tool-result cleanup that runs afterwards can empty the conversation. The real-world trigger is a **corrupt conversation** — typically an empty upstream response recorded as a complete tool message — not a large `CLAUDE.md`.

This does not soften the defect: the leak was real whenever it fired. But it invalidated every fixture, and it means the user-facing text was misdescribing the failure as well as leaking. **Both pre-existing tests for this path never executed the line they were named for.** They passed without ever reaching it.

## A second, unguarded site

`_apply_compaction` runs pairing validation *again* after `_compact_messages` returns, and `_compact_messages` short-circuits below the compaction threshold. So a below-threshold conversation could be emptied with no post-condition anywhere, and a system-only body went upstream. Closed here.

Both guards are scoped to **what the bridge itself removed**. A request that arrives unsendable — `messages: []`, or system-only — is the provider's to reject, not ours. An earlier draft of this fix got that wrong and rejected every `messages: []` request; the full suite caught it, and there is now a test pinning it.

## Design decisions worth reviewing

- **Raise, don't return a sentinel.** The defect was a value flowing to the upstream call. A sentinel can be ignored by a future call site; an exception cannot. Cost: **nine `except CompactionFailedError` clauses** — eight in the handlers, plus the fail-over catch in `_request_with_retry_balancing`. (An earlier revision of this description said "eight", counting only the handlers; corrected after the automated reviewer flagged the undercount.)
- **Eight handler clauses, not four.** The Messages and Responses handlers have an outer `try` that covers the pre-flight site, but their **inner** `except Exception` around `_request_with_retry` shadows it on the recovery path, so each handler needs its own clause at both sites.
- **`error.reason == "compaction_failed"` is a test-observability requirement, not decoration.** `_check_request_size` returns a byte-identical 400 envelope from the very next line of every handler. Without the marker, no acceptance assertion can tell the two apart — and every such test would pass against unfixed code. `test_size_rejection_is_not_reported_as_compaction_failure` is the negative control.
- **The recovery path fails over without cooling backends down.** No second upstream request was made, so there is no evidence against that backend; a cooldown would 503 every concurrent session because one conversation was corrupt. This matches the policy the surrounding code already states. The `compaction_exhausted` flag resets per attempt, so an irreducible conversation on backend A cannot be reported as the verdict for backend B's rate limit — which would tell the user not to retry the one thing that would work.
- **Streaming gets a plain HTTP 400, not an SSE error event.** Compaction runs *before* the stream branch in all four handlers, so nothing has been written. This matches Anthropic's own API, which reserves `event: error` for failures after a 200. Worth noting: streaming is the dominant path, so this is not an edge case.

## Evidence the tests are real

Every new test was run against the base revision and observed to fail for the right reason before the fix was applied.

| Group | At base revision | With the fix |
|---|---|---|
| `test_compaction.py` | 8 failed — 5 "DID NOT RAISE", 3 printing the leaked `[Kitty Bridge: …]` string | 12 pass |
| `test_compaction_failure_response.py` | 13 failed — the headline one returning **HTTP 200**: the request succeeded and carried the vendor string upstream | 15 pass |
| `test_vendor_token_guard.py` | 3 failed | 9 pass |

The HTTP 200 is the end-to-end proof the defect was live, not theoretical.

Two of the new tests correctly pass at base: the negative control and the sticky-flag regression guard. Both protect the *new* code's correctness rather than the original defect.

## Local gates

**CI is green on all four Python versions**, plus CodeQL, both static analyses, the automated `review` job and `ci-required`.

Locally: `ruff check .`, `lint-imports`, `mypy src/kitty` clean. Two suite failures reproduce on an unmodified tree, in files this PR does not touch — `tests/test_egress_https_proxy.py` (the local `openssl` rejects `-copy_extensions`, so the test CA cannot be generated) and one IPv6 address-mapping case in `test_bridge_management.py`. Both pass in CI.

The handler suite runs in **1.18s**, down from 60.69s: the automated reviewer caught that one regression guard was falling into the empty-response retry ladder and sleeping 20s + 40s per run — about four minutes per PR across the version matrix, for an assertion that takes milliseconds.

## Design documents

- Register row **M13 withdrawn** — it no longer mutates anything.
- **F3 marked fixed**, with both corrections recorded so the wrong trigger cannot be reintroduced.
- **G14 closed**; **Q9 answered** and recorded (return downstream only; the message names the product, since it never leaves the user's machine).
- **TR-4's acceptance exemption lifted** — that assertion now gates for real.
- **T-G5 (KBR-81) note:** this fix removes the last live bridge-introduced vendor string, so that guard loses its only natural positive fixture. It inherits the synthetic historical string kept in `test_vendor_token_guard.py`.

## Scope deliberately not taken

- The full per-adapter vendor-token guard needs the projection oracle — that is T-G5 (KBR-81). The guard here scans **source literals** against an allowlist, never traffic, so it avoids the flat-scan trap that would put fidelity and indistinguishability in conflict.
- Acceptance scenario TR-4 belongs to T-J2 (KBR-108).
- **A residual hole worth its own ticket:** streaming requests have no tighter-budget recovery from an upstream context-too-large at all. Pre-existing, recorded, not addressed here.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[KBR-5]: https://shelpuk.atlassian.net/browse/KBR-5?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by shelpuk** [comment]
## Review round 1 addressed (`78c4dae`), plus one CI note that is not from this PR

### Three findings, all real

**1. Recovery-path coverage was one handler out of four.** Only `/v1/chat/completions` had a recovery-path test. The other three catch the exception at a *different* site from the pre-flight one, and each degrades differently without it — Messages and Responses through their inner `except Exception` around `_request_with_retry`, Gemini and Chat Completions through theirs, and the two lacking an outer `try` through the access-log middleware. A regression removing or misordering any of those three clauses would have gone unnoticed. Now covered per handler.

That one existing test also monkeypatched `_is_oversized_request` to return `True`, sidestepping the 600 KB gate it was meant to prove. Added a test that sends a real ~700 KB history through the real gate.

**2. Nothing asserted what actually leaves the process.** Every other test here asserts that *nothing* is sent. `TestNothingOnTheWireNamesTheProduct` now lets a normal request through to a recording upstream and scans both the body and the header set — the only coverage of the header channel in the suite. It doubles as the fidelity complement: the inbound turn deliberately reads `Please explain how kitty-bridge works` and must arrive **byte-identically**, so the guard can never be satisfied by stripping the user's own words.

**3. A real race in the failover flag.** `compaction_exhausted` was reset at the top of each attempt — *ahead of* the `break` taken when the pool goes unhealthy. An attempt that never started would clear what the previous one established, silently turning a clear compaction error back into the raw upstream error. The reset now happens after that exit.

Also added the `Args`/`Returns` sections `_is_oversized_request` was missing.

### One finding that changed a claim in this PR

Building the real-payload fixture, it would not trigger. Rather than force it, I tested why: **the recovery-path post-condition is not reachable with realistic input.**

Pre-flight `_apply_compaction` strips every orphan tool result (raising if that empties the conversation), so anything arriving at the recovery re-compaction is already well-paired — and `_compact_messages` groups `tool_use`/`tool_result` atomically and its guaranteed-fit fallback always keeps one non-system block. Four oversized shapes were driven through both stages; none reached it.

The guard stays, for two reasons: `_compact_with_tighter_budget` does **not** re-run pairing validation, so a future change could open the path, and the invariant should hold wherever compaction runs. But the handler tests for that site now **state that they inject the exception**, rather than implying a fixture provokes it, and the finding is recorded in `TEST_SUITE.md` under F3. A test that quietly proves nothing is precisely what let this defect live in the first place.

### CI: `review-scripts` fails, and it is not from this PR

`test / Python 3.10`, `3.11`, `3.12`, `3.13`, `CodeQL`, `Analyze (actions)`, `Analyze (python)` and `review` all **pass**. The failure is `review-scripts`, which `ci-required` gates on.

It is pre-existing — **`main` is already red from it**, and this branch touches neither `.gitignore` nor `.github/review/`.

Cause: `GitignoredDocumentsAreNotPromisedTests::test_the_imperative_rule_recognises_what_it_forbids` asserts `.system_design` is among the gitignored prefixes. That was true until #43 (KBR-2) deliberately un-ignored `.system_design/` so the design docs could be tracked. The assertion is a positive control for the sweep above it — it is doing its job, telling us its own fixture is now stale, but its premise no longer holds.

The fix is one line — repoint the control at a path that is still ignored, e.g. `.requirements` — but it is CI infrastructure unrelated to this defect, so I have deliberately **not** folded it into a priority-0 fix. Happy to open it as its own PR on request. It will block this one from going green until it lands.

Two further failures in that job (`RetryGateWiringTests::test_an_unusable_stamp_is_unmeasured_rather_than_enormous`) also reproduce on an unmodified tree.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by shelpuk** [comment]
## Added to scope at the product owner's request: fixing `main` (`16782c7`)

`review-scripts` has been red on `main` since #43. Folding the fix in here rather than shipping it separately.

### The one-line cause, and the larger thing behind it

#43 un-ignored `.system_design/` so the design documents could be tracked. A positive control in the review system's own tests still asserted that directory was gitignored, so it failed — and took the whole job, and therefore `ci-required`, with it. The control was working correctly; its fixture had gone stale.

That was the smaller half. The review system told the reviewer, in five separate places, that `.system_design/` is **not** in the checkout:

> Do not look for a design document, do not report a missing one as a finding, and do not claim to have read one.

That has been false since #43. On a repository whose design documents *are* the internal specification, every automated review since then has been explicitly instructed to ignore the specification it was standing on. Fixing only the assertion would have turned CI green while leaving that in place.

### What changed

- **`REVIEW_PROMPT.md`, `REVIEW_GUIDE.md`, `README.md`, `rules/docs.md`, `rules/packaging.md`** — `.system_design/` is tracked and should be read for the area a change touches. `README.md` is the user-facing interface contract; the design documents are the internal one. `.requirements/` and `CLAUDE.md` are still excluded and still must not be promised to the reviewer. The guidance says to read the relevant section rather than the whole file — `TEST_SUITE.md` alone is 133 KB and the prompt budget is real.
- **`rules/docs.md`** names the failure that gets missed. Not just code contradicting a stable design, but **a design left stale by a change that invalidates it** — a wrong-docstring defect one level up, and worse in one respect, because the next author reads it as the specification.
- **`rules/packaging.md`** makes re-ignoring `.system_design/` a critical finding. Nothing else would catch that regression: the reviewer would simply go quiet, and a quiet reviewer looks exactly like a clean diff.
- **The control** now derives its specimens from what `.gitignore` actually says instead of hard-coding a path, so un-ignoring a directory can never take the review job red again — while still failing loudly if the parser stops reading `.gitignore` at all. It now asserts `.system_design` is **not** ignored, which is the claim that matters today.
- **`select_rules.py`** needed no selector change: it already matched `.system_design/**/*.md` before the directory was un-ignored, which was foresight worth keeping. Only its comments were wrong, claiming those paths were unreachable.

### On diverging from upstream

These files are carried from the repository this review system was adopted from, and `.github/review/rules/ci.md` requires a stated reason for a divergence. The reason: the sentence is false in this repository and silently degrades every review run here. `select_rules.py` is already the designated local file.

### Verification

`.github/review/tests/test_review_scripts.py`: **605 passed**, up from 604 with 1 failure. The repo's own `.github`-policing tests — `test_github_actions.py`, `test_review_workflow_cli_contract.py`, `test_pypi_packaging.py` — 79 passed. `ruff`, `mypy`, `lint-imports` all clean.

One test fails on my Windows checkout only (`RetryGateWiringTests`, a Git Bash `date` arithmetic difference). It passes on CI — the failing run's log shows `Ran 606 tests ... FAILED (failures=1)`, and that one failure was the `.system_design` assertion now fixed. Untouched deliberately: it is not a CI failure and not this PR's business.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: tests/bridge/test_compaction_failure_response.py [hunk 1]
```diff
@@ -0,0 +1,694 @@
...
+            raise UpstreamError(429, "rate limited")
+
+        monkeypatch.setattr(server, "_make_upstream_request", _fail)
+        monkeypatch.setattr(server, "_is_oversized_request", lambda _req: True)
+
+        def _tighter(cc_request, factor=0.5):
+            if len(calls) == 1:
+                server._compact_messages(list(_ORPHAN_CC_MESSAGES), 10)
+
+        monkeypatch.setattr(server, "_compact_with_tighter_budget", _tighter)
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply tests/bridge/test_compaction_failure_response.py:476]: Fixed in 691e974. Correct in all four rounds it was raised — thank you for holding it rather than dropping it.

The trace was right in the part I would most likely have got wrong by inspection: backend 0 stays healthy *because* the recovery path deliberately does not cool it down, the flag resets on attempt 2, so the attempt loop ends with nothing to raise and control drops through to the ladder.

Applied the same one-line stub the sibling test already had, with the reason recorded inline so it does not later get removed as noise.

The cost was larger than the estimate: the whole module went from **60.69s to 1.18s**, because that single sleep dominated it. Across four Python versions that is roughly four minutes per pull request run.

File: tests/bridge/test_compaction_failure_response.py [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+
+
+class TestErrorMessageIsActionable:
+    """The message is for the user, so it has to be true and useful.
+
+    The text it replaced blamed the system prompt and told the user to run
+    ``/clear``. Neither is reliably right: a large system prompt cannot cause
+    this, and the real cause is a lost tool-call pairing.
+    """
+
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply tests/bridge/test_compaction_failure_response.py:682]: Fixed in 691e974 — this is the same finding as the round-4 thread, where the full response is.

One note on the alternative you offered: I took the `_EMPTY_FINAL_DELAYS` stub but **not** the stubbed `_select_backend` that raises `AllBackendsUnhealthyError`. Forcing that exception would have replaced the behaviour under test with a fixture — the point of this guard is that the loop reaches its natural end with `compaction_exhausted` correctly reset, and a stubbed selector would have made it pass whether or not the reset works. Emptying the delay list leaves every branch intact and only removes the waiting.

Module runtime: 60.69s → 1.18s.

File: src/kitty/bridge/server.py [hunk 3]
```diff
@@ -1825,6 +1844,67 @@ def _attribution_headers(self) -> dict[str, str]:
...
+    #: wrongly; the guaranteed-fit fallback always keeps a non-system block, so
+    #: a large system prompt alone cannot cause this.
+    _COMPACTION_FAILED_MESSAGE = (
+        "Kitty Bridge could not reduce this conversation to something the model can accept: "
+        "after compaction no messages were left to send. This usually means the conversation "
+        "contains a tool result whose matching tool call was lost. Start a new conversation "
+        "(/clear), or switch to a model with a larger context window."
+    )
+
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (shelpuk) [reply src/kitty/bridge/server.py:1865]: Addressed in 78c4dae, and the round-5 review confirms it — the message now leads with `/clear` and offers the larger context window second.

Taking the reasoning further than the wording, because it is the more useful half: you are right that the two remedies are not symmetric. For the real trigger — an orphan tool result, i.e. a conversation that cannot be parsed — a larger context window does nothing at all. The current text reads:

> Start a new conversation (/clear), or switch to a model with a larger context window.

I stopped short of `"; if the issue persists, ..."`. `/clear` fixes the dominant case outright, so "if it persists" would describe a situation that should not arise, and inviting a user to go and change models after a `/clear` that already worked is worse advice than the flat "or". If we later learn of a case where `/clear` genuinely does not resolve it, that sentence becomes right and I would take it then.

Recorded rather than argued silently: the assertion in `test_message_names_the_product_and_the_real_cause` deliberately pins the *content* (product name, the tool-pairing cause, both remedies) and not the ordering, so a future copy edit is not blocked by a test — but `test_message_does_not_blame_the_system_prompt` does pin the one claim that was measurably false before.

**Comment by shelpuk** [comment]
## Review round 5 addressed (`691e974`) — all checks green

**CI is fully green for the first time:** all four Python versions, `CodeQL`, both static analyses, `review`, `review-scripts` and **`ci-required`**.

### The standing warning — fixed, and it was worse than estimated

`test_a_later_unrelated_failure_is_reported_as_itself` was falling into the `_EMPTY_FINAL_DELAYS` ladder. The reviewer's trace was exactly right, including the part I would have got wrong by inspection: backend 0 stays healthy *because* the recovery path deliberately does not cool it down, the `compaction_exhausted` flag resets on attempt 2, so the attempt loop ends with nothing to raise and drops through to the ladder — 20s then 40s of real sleep before the 429 under test surfaces.

Same one-line stub the sibling test already had, with the reason recorded inline so it does not get removed later as noise.

**Module runtime: 60.69s → 1.18s.** That single sleep was dominating the whole file, so the saving is closer to four minutes per PR across the version matrix than the one minute the finding claimed.

Raised in four consecutive rounds before I acted on it. That is on me, not the reviewer — it was correct the first time.

**One part of the suggestion I did not take.** Round 1 also offered a stubbed `_select_backend` that raises `AllBackendsUnhealthyError`. Forcing that would have replaced the behaviour under test with a fixture: the point of this guard is that the loop reaches its natural end with `compaction_exhausted` correctly reset, and a stubbed selector would make it pass whether or not the reset works. Emptying the delay list leaves every branch intact and removes only the waiting.

### Message ordering — addressed earlier, reasoning recorded

Fixed in `78c4dae`; this round's review confirms `/clear` now leads. I stopped short of the suggested `"; if the issue persists, …"` and said why on the thread: `/clear` resolves the dominant case outright, so inviting a user to go and change models afterwards would be worse advice than the flat "or". If a case turns up where `/clear` genuinely does not resolve it, that sentence becomes correct and I will take it.

### Description inaccuracy — corrected

The reviewer flagged, correctly, that the description said "eight `except` clauses" when the real count is **nine**. Verified: eight in the four handlers (two each) plus the fail-over catch in `_request_with_retry_balancing`, with three `raise` sites. My "eight" was counting only the handlers and read as a total. The description now says nine and names the ninth.

Also refreshed two stale sections: the local-gates paragraph (CI is green now, and the two local-only failures are named as environmental), and the removal of the `main`-fix section, since that work was reverted out of this PR at the product owner's request — `main` was fixed independently in `67fd7aa`, whose commit message scopes the remaining reviewer-documentation change to separate, already-tracked work.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by shelpuk** [comment]
## Review round 6 addressed (`5c31332`)

### The new finding was right, and it is the one I should have caught myself

`README.md` — the new user-facing 400 was not in Troubleshooting. This change introduces an error a user will read and then search for, and the repo's own documentation rule is explicit that behaviour moving without the README is drift. The review guide even lists "an error message a user will search for → *Troubleshooting* / *FAQ*". I updated four design-document sections and missed the one file a user actually reads.

The entry now sits beside the existing context-window one and says the three things someone hitting this needs:

- **nothing was sent to the provider** — worth stating, because the error arrives where an upstream failure normally would;
- the usual cause is a tool result whose matching tool call was lost, so **a larger context window does not help** and `/clear` does;
- if it recurs immediately on a fresh conversation, the system prompt and tool definitions are the thing to trim.

It also names `"reason": "compaction_failed"` so the response is distinguishable in a log from the ordinary request-too-large rejection.

**Guarded, so it cannot drift apart again.** `test_the_readme_documents_this_error` anchors on the message's first clause rather than the whole string — rewording the tail stays free, changing what the user actually searches for does not. Verified as a real guard rather than decoration: with the README entry removed it fails, and the failure names the missing text.

### The description undercount

Round 6 re-raised the "eight vs nine `except` clauses" inaccuracy. That was already corrected in the description before this round submitted — the review job started on the push of `691e974` and read the description as it stood then. Verified independently: nine clauses at server.py:2559, 2578, 3128, 3141, 4114, 4135, 4710, 4894, 4907, with three `raise` sites at 4794, 5819, 6076. Eight are the handlers (two each), the ninth is the fail-over catch in `_request_with_retry_balancing`.

### Standing findings

Both are closed and confirmed by this round: the `_EMPTY_FINAL_DELAYS` sleep (`691e974`, module 60.69s → 1.18s) and the message ordering (`78c4dae`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by shelpuk** [comment]
## Round 7: no findings — and one withheld suggestion answered rather than left hanging

Round 7 reports **no findings**. Every warning raised across the seven rounds is closed:

| Round | Finding | Resolution |
|---|---|---|
| 1 | `_COMPACTION_FAILED_MESSAGE` mis-orders the remedies | `78c4dae` — `/clear` leads |
| 1–5 | Recovery-path test sleeps ~60s per CI version | `691e974` — module 60.69s → 1.18s |
| 2–6 | Description undercounts the `except` clauses as eight | Corrected to nine; verified in code |
| 6 | New user-facing 400 missing from README Troubleshooting | `5c31332` — entry added, plus a guard against re-drift |

### The one suggestion withheld this round, answered with evidence

The round-7 prose notes a "docstring convention drift on `_apply_compaction` — added `Args:`/`Raises:` but no `Returns:`, while the in-place return type is `None`." It was withheld as suggestion-level, but it would only come back later, so: **I checked, and I am not going to apply it.**

I parsed every `-> None` function in `src/kitty` that carries a multi-line docstring:

```
-> None WITH    Returns:  0
-> None WITHOUT Returns: 60
```

**Sixty out of sixty omit it. None include it.** Adding `Returns:` to `_apply_compaction` would make it the only exception in the package — the drift would be the fix, not the current state. It also matches Google style, which omits the section for functions returning `None`.

Recording it here rather than silently declining, so the next reviewer — human or automated — does not have to re-derive it.

### Status

All checks green: four Python versions, `CodeQL`, both static analyses, `review`, `review-scripts`, `ci-required`.

The reviewer has stopped surfacing actionable feedback. Ready for human review; not merged.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
